src/api/flow.ts

```diff
@@ -117,53 +117,58 @@ export function createFlowGenerator(name: string, generator: Function) {
-                    gen,
-                    res
-                )
-            } catch (e) {
-                return rejector(e)
+        const res = new Promise(function(resolve, reject) {
+            let stepId = 0
+            rejector = reject
+
+            function onFulfilled(res: any) {
+                let ret
+                try {
+                    ret = action(`${name} - runid: ${runId} - yield ${stepId++}`, gen.next).call(
+                        gen,
+                        res
+                    )
+                } catch (e) {
+                    return reject(e)
+                }
+                next(ret)
+                return null
             }
-            next(ret)
-            return null
-        }
 
-        function onRejected(err: any) {
-            let ret
-            try {
-                ret = action(`${name} - runid: ${runId} - yield ${stepId++}`, gen.throw).call(
-                    gen,
-                    err
-                )
-            } catch (e) {
-                return rejector(e)
+            function onRejected(err: any) {
+                let ret
+                try {
+                    ret = action(`${name} - runid: ${runId} - yield ${stepId++}`, gen.throw).call(
+                        gen,
+                        err
+                    )
+                } catch (e) {
+                    return reject(e)
+                }
+                next(ret)
             }
-            next(ret)
-        }
 
-        function next(ret: any) {
-            if (ret.done) return resolver(ret.value)
-            // TODO: support more type of values? See https://github.com/tj/co/blob/249bbdc72da24ae44076afd716349d2089b31c4c/index.js#L100
-            if (!ret.value || typeof ret.value.then !== "function")
-                return fail("Only promises can be yielded to asyncAction, got: " + ret)
-            return ret.value.then(onFulfilled, onRejected)
-        }
+            function next(ret: any) {
+                if (ret.done) return resolve(ret.value)
+                // TODO: support more type of values? See https://github.com/tj/co/blob/249bbdc72da24ae44076afd716349d2089b31c4c/index.js#L100
+                if (!ret.value || typeof ret.value.then !== "function")
+                    return fail("Only promises can be yielded to asyncAction, got: " + ret)
+                return ret.value.then(onFulfilled, onRejected)
+            }
 
-        const res = new Promise(function(resolve, reject) {
-            resolver = resolve
-            rejector = reject
             onFulfilled(undefined) // kick off the process
         }) as any
-        res.cancel = function() {
-            onRejected(new Error("FLOW_CANCELLED"))
-        }
+
+        res.cancel = action(`${name} - runid: ${runId} - cancel`, function() {
+            try {
+                gen.return()
+                rejector(new Error("FLOW_CANCELLED"))
```

Comment:
So externally we're rejecting but internally we're doing cancellation.

What happens if `.cancel` is called while a flow is being `yield`ed in a flow? Preferably, the internal flow should be cancelled as well (so that, if the current promise I'm waiting for has a `.cancel` and `.cancel` was called on me - call it on that promise).

Reply:
good point, will call `cancel` on any pending promise if possible